processing/src/main/java/org/apache/druid/segment/join/table/RowBasedIndexBuilder.java

```diff
@@ -62,6 +63,11 @@ public RowBasedIndexBuilder(ColumnType keyType)
   {
     this.keyType = keyType;
 
+    // Cannot build index on complex types, and non-primitive arrays
+    if (keyType.is(ValueType.COMPLEX) || keyType.isArray() && !keyType.isPrimitiveArray()) {
```

Comment:
I think we should have this check on the planning side as well. 
That can be a follow up PR.